ocaml-lsp-server/src/dune.ml

```diff
@@ -1,46 +1,158 @@
+module Chan = struct
+  open Fiber.O
+
+  type t =
+    { in_thread : Scheduler.thread
+    ; out_thread : Scheduler.thread
+    ; out_chan : out_channel
+    ; in_chan : in_channel
+    }
+
+  let write t sexp =
+    let+ res =
+      Scheduler.async_exn t.out_thread (fun () ->
+          match sexp with
+          | None -> close_out_noerr t.out_chan
+          | Some sexp -> Csexp.to_channel t.out_chan sexp)
+      |> Scheduler.await_no_cancel
+    in
+    match res with
+    | Error e -> Exn_with_backtrace.reraise e
+    | Ok s -> s
+
+  let read t =
+    let+ res =
+      Scheduler.async_exn t.in_thread (fun () ->
+          match Csexp.input_opt t.in_chan with
+          | Error _
+          | Ok None ->
+            None
+          | Ok (Some s) -> Some s)
+      |> Scheduler.await_no_cancel
+    in
+    match res with
+    | Error e -> Exn_with_backtrace.reraise e
+    | Ok s -> s
+
+  let create scheduler in_fd out_fd =
+    let in_thread = Scheduler.create_thread scheduler in
+    let out_thread = Scheduler.create_thread scheduler in
+    let in_chan = Unix.in_channel_of_descr in_fd in
+    let out_chan = Unix.out_channel_of_descr out_fd in
+    { in_thread; out_thread; in_chan; out_chan }
+end
+
+module Drpc = Dune_rpc.V1
+
+module Client =
+  Drpc.Client
+    (struct
+      include Fiber
+
+      let parallel_iter t ~f =
+        let stream = Fiber.Stream.In.create t in
+        Fiber.Stream.In.parallel_iter stream ~f
+    end)
+    (Chan)
+
 type state =
   | Binary_not_found
   | Out_of_date
-  | Running
 
 type t =
   { scheduler : Scheduler.t
-  ; state : state Fiber.Ivar.t
+  ; diagnostics : Diagnostics.t
   }
 
-let create scheduler = { scheduler; state = Fiber.Ivar.create () }
+let create scheduler diagnostics = { scheduler; diagnostics }
 
-let state t : state Fiber.t =
-  let open Fiber.O in
-  let* state = Fiber.Ivar.peek t.state in
-  match state with
-  | Some s -> Fiber.return s
-  | None -> (
-    match Bin.which "dune" with
-    | None ->
-      let res = Binary_not_found in
-      let+ () = Fiber.Ivar.fill t.state res in
-      res
-    | Some bin ->
-      let bin = Fpath.to_string bin in
-      let dev_null () = Unix.openfile "/dev/null" [ Unix.O_RDWR ] 0o666 in
-      let stdin = dev_null () in
-      let stdout = dev_null () in
-      let stderr = dev_null () in
+let lsp_of_dune dune =
+  let module D = Drpc.Diagnostic in
+  let range =
+    match D.loc dune with
+    | None -> Range.first_line
+    | Some loc ->
+      let loc =
+        { Loc.loc_start = loc.start; loc_end = loc.stop; loc_ghost = false }
+      in
+      Range.of_loc loc
+  in
+  let severity =
+    D.severity dune
+    |> Option.map ~f:(function
+         | D.Error -> DiagnosticSeverity.Error
+         | Warning -> DiagnosticSeverity.Warning)
+  in
+  let message = Format.asprintf "%a@." Pp.to_fmt (D.message dune) in
+  Diagnostic.create ~range ?severity ~source:"dune" ~message ()
 
+let run_rpc t bin =
+  let stdin_i, stdin_o = Unix.pipe () in
```

Comment:
I don't see these being closed. Should `Chan` be responsible for closing the fd properly? If so, it could also be responsible to create these and provide a function `val fd: t -> (Unix.fd * Unix.fd)`

Reply:
We don't close the read end because we never stop reading I believe. The write end is closed when we close the channel (see `Chan.write`).

> If so, it could also be responsible to create these and provide a function val fd: t -> (Unix.fd * Unix.fd)

The thing is, it's only responsible 1 read and 1 write end of the pipe. So it's error prone that it needs to return fd's that will have to be closed by the spawned process. That's why I opted to just create them upfront.

Reply:
> We don't close the read end because we never stop reading I believe.

I remember a similar statement in another PR. That's fine if we assume `dune init` will never crash, but it's a bold assumption. Granted dune does not crash often, but it happens, and if it does at the moment, I think the only workaround will be to restart the language server, right? Just playing devil's advocate, I don't know how much effort we want to put in being fault-tolerant against dune.

Reply:
Yes, that's a  good point. I've changed the code to work as follows:

* dune rpc init crashes
* we are unable to read from the read end of the pipe
* we now close the read of the pipe as a result